### ISSUE TITLE
[TW-644] Circular dependency detection code incorrect

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#679]
_Steve Kroon on 2011-12-08T13:59:35Z says:_ 

 See discussion and examples at [[http://taskwarrior.org/boards/1/topics/1754]].  Essentially, the code detects non-circular situations as circular, and moreover only detects those non-circular situations it wishes to complain about when additional dependencies are introduced, rather than when the (non-)problem is introduced.

I have tested this in 1.9.4, and the dependency handling code in 2.0.0.beta4 looks the same.  The problem is in the approach followed in followUpstream of dependency.cpp .  I believe a better approach when adding a dependency of A on B, is to do a depth-first-search on the dependency subgraph generated by following dependency connections starting with A->B, and looking for back edges.  The current solution fails because it incorrectly perceives cross edges in a DFS as causing circularity.

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2011-12-08T13:59:35Z
Modified: 2014-02-09T02:06:31Z
```

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-12-10T16:06:48Z says:_ 

 Thank you Steve - this is well researched.  Much appreciated.

**Comment by taskwarrior** [comment]
_Wim Schuermann on 2011-12-19T11:53:20Z says:_ 

 Since the corresponding forum thread is still seeing activity and people seem to be waiting for a fix, let me just mention that this bug has been fixed in the 2.0.0 branch by commit 726fc33; this issue will probably stay open until proper unit tests have been added.
